futures-util/src/lock/rwlock.rs

```diff
@@ -0,0 +1,462 @@
+#![allow(dead_code)]
+#![allow(unused_imports)]
+#![allow(unused_mut)]
+#![allow(unused_variables)]
+
+use futures_core::future::{FusedFuture, Future};
+use futures_core::task::{Context, Poll};
+use slab::Slab;
+use std::cell::UnsafeCell;
+use std::fmt;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Mutex;
+use std::task::Waker;
+
+// Sentinel for when no slot in the `Slab` has been dedicated to this object.
+const WAIT_KEY_NONE: usize = usize::max_value();
+
+const PHASE: usize = 1;
+const LOCKED: usize = 1 << 1;
+const ONE_READER: usize = 1 << 2;
+const READER_OFFSET: usize = 2;
+const READER_OVERFLOW: usize = 1 << (usize::BITS - 1);
+
+enum Ticket {
+    Stale(usize),
+    None,
+}
+
+enum Waiter {
+    Waiting(Waker),
+    Woken,
+}
+
+impl Waiter {
+    fn register(&mut self, waker: &Waker) {
+        match self {
+            Waiter::Waiting(w) if waker.will_wake(w) => {}
+            _ => *self = Waiter::Waiting(waker.clone()),
+        }
+    }
+
+    fn wake(&mut self) -> bool {
+        match mem::replace(self, Waiter::Woken) {
+            Waiter::Waiting(waker) => waker.wake(),
+            Waiter::Woken => return false,
+        }
+
+        true
+    }
+}
+
+struct WaiterSet {
+    waiters: Mutex<Slab<Waiter>>,
+}
+
+impl WaiterSet {
+    fn new() -> Self {
+        WaiterSet { waiters: Mutex::new(Slab::new()) }
+    }
+
+    fn register(&self, key: usize, waker: &Waker) -> usize {
+        let mut this = self.waiters.lock().unwrap();
+
+        if key == WAIT_KEY_NONE {
+            this.insert(Waiter::Waiting(waker.clone()))
+        } else {
+            this[key].register(waker);
+            key
+        }
+    }
+
+    fn remove(&self, key: usize) {
+        let mut this = self.waiters.lock().unwrap();
+
+        if key != WAIT_KEY_NONE {
+            this.remove(key);
+        }
+    }
+
+    fn notify_all(&self) -> bool {
+        let mut this = self.waiters.lock().unwrap();
+        this.iter_mut().fold(false, |acc, (_, waiter)| waiter.wake() || acc)
+    }
+
+    fn notify_one(&self) -> bool {
+        let mut this = self.waiters.lock().unwrap();
+        this.iter_mut().any(|(_, waiter)| waiter.wake())
+    }
+}
+
+/// A futures-aware read-write lock.
+pub struct RwLock<T: ?Sized> {
+    state: AtomicUsize,
+    blocked: WaiterSet,
+    writers: WaiterSet,
+    value: UnsafeCell<T>,
```

Comment:
If we use the `spin` crate, and replace the `AtomicUsize/UnsafeCell` pair with a `spin::RwLock` we may deduplicate a lot of code.